src/transformers/models/mot/modeling_mot.py

```diff
@@ -0,0 +1,1621 @@
+
+        attn_output = self._merge_heads(attn_output, self.num_heads, self.head_dim)
+        attn_output = self.c_proj(attn_output)
+        attn_output = self.resid_dropout(attn_output)
+
+        outputs = (attn_output, present)
+        if output_attentions:
+            outputs += (attn_weights,)
+
+        return outputs  # a, present, (attentions)
+
+
+class MoTMLP(nn.Module):
+    r"""
+    Implementation of the Mixture of Tokens Sparse MLP module.
+    """
+
+    def __init__(self, inner_dim: int, config: MoTConfig, sparsity_dim: int = 0, init_type: str = "kaiming_uniform"):
+        super().__init__()
+
+        self.d_model: int = config.n_embd
+        self.d_ff: int = config.n_inner if inner_dim is None else inner_dim
+        self.n_expert: int = config.n_expert
+        self.group_size: int = config.group_size
+        self.sparsity_dim: int = sparsity_dim
+        self.expert_size: int = config.expert_size
+        self.temperature: float = config.temperature
+        self.act = ACT2FN[config.activation_function]
+
+        self.init_type: str = init_type
+        self.init_scale: float = config.init_scale
+
+        self.emit_softmax_over_experts: bool = config.emit_softmax_over_experts
+        self.use_discrete_routing: bool = config.use_discrete_routing
+
+        if self.n_expert is not None:
+            if self.d_ff % self.n_expert:
+                self.d_ff += self.n_expert - (self.d_ff % self.n_expert)
+                warnings.warn("d_ff should be divisible by n_expert, padding d_ff to be divisible by n_expert")
+            self.expert_size = self.d_ff // self.n_expert
+        elif self.expert_size is not None:
+            if self.d_ff % self.expert_size:
+                self.d_ff += self.expert_size - (self.d_ff % self.expert_size)
+                warnings.warn("d_ff should be divisible by expert_size, padding d_ff to be divisible by expert_size")
+            self.n_expert = self.d_ff // self.expert_size
+        else:
+            raise ValueError("Either expert_size or n_expert should be provided")
+
+        self.lin1 = nn.Parameter(
+            self.get_init_weight(
+                (self.n_expert, self.d_model, self.expert_size),
+                fan_in=self.d_model,
+                init_type=self.init_type,
+                scale=self.init_scale,
+            )
+        )
+
+        self.lin2 = nn.Parameter(
+            self.get_init_weight(
+                (self.n_expert, self.expert_size, self.d_model),
+                fan_in=self.expert_size,
+                init_type=self.init_type,
+                scale=self.init_scale,
+            )
+        )
+
+        self.controller = nn.Parameter(
+            self.get_init_weight(
+                (self.d_model, self.n_expert),
+                fan_in=self.d_model,
+                init_type=self.init_type,
+                scale=self.init_scale,
+            )
+        )
```

Comment:
we never call init weights in the init of modules, but in the `PreTrainedModel` class !